contracts/vortex/CarbonVortex.sol

```diff
@@ -138,109 +345,502 @@ contract CarbonVortex is ICarbonVortex, Upgradeable, ReentrancyGuardUpgradeable,
+        // allocate array for the fee amounts for the tokens
+        uint256[] memory feeAmounts = new uint256[](len);
         // allocate array for the reward amounts for caller
         uint256[] memory rewardAmounts = new uint256[](len);
         // cache rewardsPPM to save gas
         uint256 rewardsPercentage = _rewardsPPM;
 
-        // withdraw fees, load balances and reward amounts
+        // withdraw fees from carbon, vault and old vortex
         for (uint256 i = 0; i < len; i = uncheckedInc(i)) {
+            Token token = tokens[i];
             // withdraw token fees
-            _carbonController.withdrawFees(tokens[i], type(uint256).max, address(this));
-            // get token balance
-            balances[i] = tokens[i].balanceOf(address(this));
+            uint256 totalFeeAmount = 0;
+            if (address(_carbonController) != address(0)) {
+                totalFeeAmount += _carbonController.withdrawFees(token, type(uint256).max, address(this));
+            }
+            if (address(_vault) != address(0)) {
+                // get vault token balance
+                uint256 vaultBalance = token.balanceOf(address(_vault));
+                // withdraw vault token balance
+                _vault.withdrawFunds(token, payable(address(this)), vaultBalance);
+                totalFeeAmount += vaultBalance;
+            }
+            if (address(_oldVortex) != address(0)) {
+                // get old vortex token balance
+                uint256 oldVortexBalance = token.balanceOf(address(_oldVortex));
+                // withdraw old vortex token balance
+                _oldVortex.withdrawFunds(token, payable(address(this)), oldVortexBalance);
+                totalFeeAmount += oldVortexBalance;
+            }
+            feeAmounts[i] = totalFeeAmount;
+
             // get reward amount for token
-            rewardAmounts[i] = MathEx.mulDivF(balances[i], rewardsPercentage, PPM_RESOLUTION);
+            rewardAmounts[i] = MathEx.mulDivF(totalFeeAmount, rewardsPercentage, PPM_RESOLUTION);
         }
 
-        // convert tokens to BNT
+        // go through all tokens and start / reset dutch auction if necessary
         for (uint256 i = 0; i < len; i = uncheckedInc(i)) {
             Token token = tokens[i];
-            // skip token if no token balance or token is BNT - no need to swap in this case
-            if (balances[i] == 0 || token == _bnt) {
+            uint256 totalFeeAmount = feeAmounts[i];
+            // skip the final target token
+            if (token == _finalTargetToken) {
                 continue;
             }
-
-            // get trade amount for token
-            uint256 tradeAmount;
-            unchecked {
-                // safe because balances[i] >= rewardAmounts[i]
-                tradeAmount = balances[i] - rewardAmounts[i];
+            // skip token if no fees have accumulated or token pair is disabled
+            if (totalFeeAmount == 0 || _disabledPairs[token]) {
```

Comment:
If a token is disabled, it would still withdraw it above and just do nothing. Perhaps a disabled token should be just ignored (or reverted) altogether?

Reply:
Tokens get disabled for trading only - it should still be withdrawable (product spec)